agents/builtin/mysql/mysql.go

```diff
@@ -93,78 +114,179 @@ func (m *MySQL) Run(ctx context.Context) {
-				m.changes <- Change{Status: inventorypb.AgentStatus_STARTING}
 				continue
 			}
 
 			if !running {
-				m.changes <- Change{Status: inventorypb.AgentStatus_RUNNING}
 				running = true
+				m.changes <- Change{Status: inventorypb.AgentStatus_RUNNING}
 			}
 
-			select {
-			case <-ctx.Done():
-				t.Stop()
-				break
-			case m.changes <- Change{Request: request}:
-				// nothing
-			}
+			m.changes <- Change{Request: &qanpb.CollectRequest{MetricsBucket: buckets}}
 		}
 	}
 }
 
-func (m *MySQL) get(q *reform.Querier) (qanpb.CollectRequest, error) {
-	var res qanpb.CollectRequest
-	structs, err := q.SelectAllFrom(eventsStatementsSummaryByDigestView, "")
+func (m *MySQL) runHistoryCacheRefresher(ctx context.Context) {
+	t := time.NewTicker(refreshHistory)
+	defer t.Stop()
+
+	for {
+		if err := m.refreshHistoryCache(); err != nil {
+			m.l.Error(err)
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			// nothing, continue loop
+		}
+	}
+}
+
+func (m *MySQL) refreshHistoryCache() error {
+	current, err := getHistory(m.db.Querier)
 	if err != nil {
-		return res, err
+		return err
+	}
+	m.historyCache.refresh(current)
+	return nil
+}
+
+func (m *MySQL) getBuckets(periodStart time.Time, periodLength time.Duration) ([]*qanpb.MetricsBucket, error) {
+	current, err := getSummaries(m.db.Querier)
+	if err != nil {
+		return nil, err
+	}
+
+	buckets := makeBuckets(current, m.summaryCache.get(), m.l)
+
+	m.summaryCache.refresh(current)
+
+	// add examples from history cache
+	history := m.historyCache.get()
+	for i, b := range buckets {
+		if esh := history[b.Queryid]; esh != nil {
+			// TODO test if we really need that
+			if b.DSchema == "" {
+				b.DSchema = pointer.GetString(esh.CurrentSchema)
+			}
+
+			if esh.SQLText != nil {
+				b.Example = *esh.SQLText
+				b.ExampleFormat = qanpb.ExampleFormat_EXAMPLE
+				b.ExampleType = qanpb.ExampleType_RANDOM
+			}
+		}
+
+		buckets[i] = b
 	}
 
-	for _, str := range structs {
-		ess := str.(*eventsStatementsSummaryByDigest)
+	startS := uint32(periodStart.Unix())
+	lengthS := uint32(periodLength.Seconds())
+	for i, b := range buckets {
+		b.PeriodStartUnixSecs = startS
+		b.PeriodLengthSecs = lengthS
+		buckets[i] = b
+	}
+
+	return buckets, nil
+}
+
+// makeBuckets uses current state of events_statements_summary_by_digest table and accumulated previous state
+// to make metrics buckets.
+//
+// makeBuckets is a pure function for easier testing.
+func makeBuckets(current, prev map[string]*eventsStatementsSummaryByDigest, l *logrus.Entry) []*qanpb.MetricsBucket {
+	res := make([]*qanpb.MetricsBucket, 0, len(current))
 
-		// skipping catch-all row
-		if ess.Digest == nil || ess.DigestText == nil {
-			m.l.Debugf("Skipping %s.", ess)
+	for digest, currentESS := range current {
+		prevESS := prev[digest]
+		if prevESS == nil {
+			prevESS = new(eventsStatementsSummaryByDigest)
+		}
+		count := float32(currentESS.CountStar - prevESS.CountStar)
+		switch {
+		case count == 0:
```

Comment:
is it possible to have the same CountStar value in prevESS and currentESS after Truncate?

Reply:
Yes, but we can't differentiate those two situations.